compiler/rustc_macros/src/diagnostics/subdiagnostic_builder.rs

```diff
@@ -0,0 +1,612 @@
+
+                    match nested_name {
+                        "code" => {
+                            let formatted_str = self.build_format(&value.value(), value.span());
+                            code.set_once((formatted_str, span));
+                        }
+                        _ => throw_invalid_nested_attr!(attr, &nested_attr, |diag| {
+                            diag.help("`code` is the only valid nested attribute")
+                        }),
+                    }
+                }
+
+                let Some((code, _)) = code else {
+                    span_err(span, "`#[suggestion_part(...)]` attribute without `code = \"...\"`")
+                        .emit();
+                    return Ok(quote! {});
+                };
+                let binding = info.binding;
+
+                Ok(quote! { suggestions.push((#binding, #code)); })
+            }
+            _ => throw_invalid_attr!(attr, &Meta::List(list), |diag| {
+                let span_attr = if let SubdiagnosticKind::MultiPartSuggestion { .. } = kind {
+                    "suggestion_part"
+                } else {
+                    "primary_span"
+                };
+                diag.help(format!(
+                    "only `{span_attr}`, `applicability` and `skip_arg` are valid field attributes",
+                ))
+            }),
+        }
+    }
+
+    pub fn into_tokens(&mut self) -> Result<TokenStream, DiagnosticDeriveError> {
+        let Some((kind, slug)) = self.identify_kind()? else {
+            throw_span_err!(
+                self.variant.ast().ident.span().unwrap(),
+                "subdiagnostic kind not specified"
+            );
+        };
+
+        let init = match &kind {
+            SubdiagnosticKind::Label
+            | SubdiagnosticKind::Note
+            | SubdiagnosticKind::Help
+            | SubdiagnosticKind::Warn
+            | SubdiagnosticKind::Suggestion { .. } => quote! {},
+            SubdiagnosticKind::MultiPartSuggestion { .. } => {
+                quote! { let mut suggestions = Vec::new(); }
+            }
+        };
+
+        let attr_args: TokenStream = self
+            .variant
+            .bindings()
+            .iter()
+            .filter(|binding| !binding.ast().attrs.is_empty())
+            .map(|binding| self.generate_field_attr_code(binding, &kind))
+            .collect();
+
+        let primary_span_field = self.primary_span_field.as_ref().map(|(span, _)| span);
+
+        let diag = &self.diag;
+        let name =
+            format_ident!("{}{}", if primary_span_field.is_some() { "span_" } else { "" }, kind);
+        let message = quote! { rustc_errors::fluent::#slug };
+        let call = match kind {
+            SubdiagnosticKind::Suggestion { suggestion_kind, code } => {
+                let applicability = self.applicability.take().map_or_else(
+                    || {
+                        span_err(self.span, "suggestion without `applicability`").emit();
```

Comment:
Do we do this in the `SessionDiagnostic` or fallback to a default?

Reply:
`SessionDiagnostic` falls back to `Unspecified`, I've changed `SessionSubdiagnostic` to do the same.